README.rst
====================
djorm-ext-pgfulltext
====================

Pgfulltext module of django orm extensions package (collection of third party plugins build in one unified package).

- Now compatible with python2 and python3 with same code base.
- Ready for django 1.3, 1.4, 1.5 and 1.6


Introduction
------------

Full Text Searching (or just text search) provides the capability to identify natural-language documents that satisfy a query, and optionally to sort them by relevance to the query. The most common type of search is to find all documents containing given query terms and return them in order of their similarity to the query. Notions of query and similarity are very flexible and depend on the specific application. The simplest search considers query as a set of words and similarity as the frequency of query words in the document. (`From postgresql documentation.`)


Classes
^^^^^^^

`djorm_pgfulltext.fields.VectorField`
    An tsvector index field which stores converted text into special format.

`djorm_pgfulltext.models.SearchManager`
    Django manager that contains helper methods for search and re/genereate indexes.


How to use it
-------------

To use it, you will need to add a new field. Obviously, this is not mandatory, as it can specify which fields you want to search at the time of calling the `search`. Keep in mind that you should put the corresponding indices for the fields to be used.

.. code-block:: python

    from djorm_pgfulltext.models import SearchManager
    from djorm_pgfulltext.fields import VectorField
    from django.db import models

    class Page(models.Model):
        name = models.CharField(max_length=200)
        description = models.TextField()

        search_index = VectorField()

        objects = SearchManager(
            fields = ('name', 'description'),
            config = 'pg_catalog.english', # this is default
            search_field = 'search_index', # this is default
            auto_update_search_field = True
        )


The manager automatically injected ``update_search_field`` method to the model instance.
Also, not to override the save method, you can pass the parameter ``auto_update_search_field = True``, so
the index field  is updated automatically by calling the ``save`` method.


Usage examples:
^^^^^^^^^^^^^^^

- The config parameter is optional and defaults to 'pg_catalog.english'.
- The fields parameter is optional. If a list of tuples, you can specify the ranking of each field, if it is None, it gets 'D' as the default.
- It can also be a simple list of fields, and the ranking will be selected by default. If the field is empty, the index was applied to all fields ``CharField`` and ``TextField``.

To search, use the ``search`` method of the manager. The current implementation, by default uses unaccent extension for ignore the accents. Also, the searches are case insensitive.

.. code-block:: python

    >>> Page.objects.search("documentation & about")
    [<Page: Page: Home page>]
    >>> Page.objects.search("about | documentation | django | home", raw=True)
    [<Page: Page: Home page>, <Page: Page: About>, <Page: Page: Navigation>]

FTS extension by default uses plainto_tsquery instead of to_tosquery, for this reason the use of raw parameter.


Update search field:
^^^^^^^^^^^^^^^^^^^^

For manual updating all search fields use management command ``update_search_field``:

.. code-block:: python

    ./manage.py update_search_field [options] appname [model]


General notes:
^^^^^^^^^^^^^^

You must ensure you have installed the extension `unaccent`:

.. code-block:: sql

    CREATE EXTENSION unaccent;
    ALTER FUNCTION unaccent(text) IMMUTABLE;

You can install this extension on template1 database for make this extension automatically available for all new created databases.

Contributing
------------

The first step of contributing is being able to run the unit tests. We provide a seet of ``docker`` containers and a ``docker-compose`` file to help you achieve this task.

Running the tests
^^^^^^^^^^^^^^^^^

The only command you need to run the tests is:

.. code-block:: bash

  docker-compose run --rm djorm python3 testing/runtests.py

Changelog
---------

**0.9.2**

- Django 1.7 lookups support.

**0.9**

- Fix django 1.6 compatibility (transaction management).


.. image:: https://d2weczhvl823v0.cloudfront.net/djangonauts/djorm-ext-pgfulltext/trend.png
   :alt: Bitdeli badge
   :target: https://bitdeli.com/free


ChangeLog
CHANGES
=======

0.9.3
-----

* Increase version number
* Add a ChangeLog file
* Add the AUTHORS file
* Update .gitignore file
* Fix flake8 issues
* Add tox.ini to the project
* Update unit tests
* Update requirements
* Update the documentation
* Add docker containers
* Update the settings file
* Package the application with PBR
* Add requirements
* Update README.rst
* Add note in the readme about searching new maintainer
* Fix default behavior of VectorField
* VectorField Defaults Should be Overrideable
* Remove .encode() from quotes method
* Update python3 string handling
* Fix django 1.8 support: using get_queryset instead of get_query_set
* Fix python3 compatibility
* Management command for updating search fields. Usage: ./manage.py update_search_field [options] appname [model] Update search fields in all models or only in specified model

0.9.2
-----

* Update version
* utf-8/unicode fixes AND tests
* ... and more utf-8 fixes
* adapt() expects utf-8 encoded strings
* Move those functions so they can be reused
* Bump version no
* Tests for Django 1.7 full text lookups
* Django 1.7 support
* Migrations & initial SQL for Django 1.7
* Renamed a few things & used psycopg2 quoting
* Pycharm & 2 virtual environments
* Support for abstract parent classes
* Support field lookups in Django 1.7
* Move GIS stuff to external module (fix #26)
* PyCharm
* Added support django gis
* Add 1.6 to the top of readme
* Update models.py
* move search method from manager onto a custom queryset class
* Add a Bitdeli badge to README
* Adding search results headlines support

0.9
---

* Increment version to 0.9
* Update README with new entry: changelog
* Update models.py
* Fix django 1.6 compatibility with transaction management
* Fix typo in models.py
* Fixed #6 - Fixed wrong behavior on autoupdate indexes (thanks to ostronom)
* Increment version to 0.7
* fix import (broken for django 1.3 and 1.4)
* Fixed #3 - Now update_search_field of model instance is more flexible
* Python3 compatiblity. (Increment version to 0.5)
* Fix typo on README
* Increment version to 0.4.2
* Fix incorrect behavior on set fields with ranking. (Fix #2)

0.4.1
-----

* Increment version to 0.4.1
* Removes unused foreign key from test models
* Fixes DatabaseError when ranking is applied with a join

0.4.0
-----

* Add license and remove versiontools dependency
* Initial commit (import from django-orm-extensions)


testing/runtests.py
# -*- coding: utf-8 -*-

import os
import sys

import django
from django.core.management import call_command


os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")

if django.VERSION[:2] >= (1, 7):
    django.setup()

if __name__ == "__main__":
    args = sys.argv[1:]
    call_command("test", *args, verbosity=2)


testing/settings.py
import os
import sys


sys.path.insert(0, '..')

PROJECT_ROOT = os.path.dirname(__file__)
DEBUG = True
TEMPLATE_DEBUG = DEBUG

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',
        'NAME': 'test',
        'USER': os.environ.get('POSTGRES_USER', 'postgres'),
        'PASSWORD': '',
        'HOST': os.environ.get('POSTGRES_HOST', 'localhost'),
        'PORT': '',
    }
}

TIME_ZONE = 'America/Chicago'
LANGUAGE_CODE = 'en-us'
ADMIN_MEDIA_PREFIX = '/static/admin/'
STATICFILES_DIRS = ()

SECRET_KEY = 'di!n($kqa3)nd%ikad#kcjpkd^uw*h%*kj=*pm7$vbo6ir7h=l'
INSTALLED_APPS = (
    'djorm_pgfulltext',
    'djorm_pgfulltext.tests',
)

TEST_RUNNER = 'django.test.simple.DjangoTestSuiteRunner'


djorm_pgfulltext/tests/models.py
# -*- coding: utf-8 -*-
from django.db import models

from ..fields import VectorField
from ..models import SearchManager

class Person(models.Model):
    name = models.CharField(max_length=32)
    description = models.TextField()
    search_index = VectorField()

    objects = SearchManager(
        fields=('name', 'description'),
        search_field = 'search_index',
        config = 'names',
    )

    def __unicode__(self):
        return self.name

    def save(self, *args, **kwargs):
        super(Person, self).save(*args, **kwargs)
        self.update_search_field()


class Person2(models.Model):
    name = models.CharField(max_length=32)
    description = models.TextField()
    search_index = VectorField()

    objects = SearchManager(
        fields=(('name', 'A'), ('description', 'B')),
        search_field = 'search_index',
        config = 'names',
    )

    def __unicode__(self):
        return self.name


class Person3(models.Model):
    name = models.CharField(max_length=32)
    description = models.TextField()
    search_index = VectorField()

    objects = SearchManager(
        fields=('name', 'description'),
        search_field = 'search_index',
        auto_update_search_field = True,
        config = 'names'
    )

    def __unicode__(self):
        return self.name


class Book(models.Model):
    author = models.ForeignKey(Person)
    name = models.CharField(max_length=32)
    search_index = VectorField()

    objects = SearchManager(
        fields=('name',),
        search_field = 'search_index',
        auto_update_search_field = True,
        config = 'names'
    )

    def __unicode__(self):
        return self.name


djorm_pgfulltext/fields.py
# -*- coding: utf-8 -*-

# Python3 string compadability
try:
    basestring
except NameError:
    basestring = str

import django
from django.db import models
from psycopg2.extensions import adapt


class VectorField(models.Field):

    def __init__(self, *args, **kwargs):
        kwargs['default'] = ''
        kwargs['editable'] = False
        kwargs['serialize'] = False
        super(VectorField, self).__init__(*args, **kwargs)

    def deconstruct(self):
        name, path, args, kwargs = super(VectorField, self).deconstruct()
        del kwargs['default']
        del kwargs['editable']
        del kwargs['serialize']
        return name, path, args, kwargs

    def db_type(self, *args, **kwargs):
        return 'tsvector'

    # def get_prep_lookup(self, lookup_type, value):
    #     if hasattr(value, 'prepare'):
    #         return value.prepare()
    #     if hasattr(value, '_prepare'):
    #         return value._prepare()
    #     raise TypeError("Field has invalid lookup: %s" % lookup_type)

    def get_db_prep_lookup(self, lookup_type, value, connection, prepared=False):
        return self.get_prep_lookup(lookup_type, value)

    def get_prep_value(self, value):
        return value

try:
    from south.modelsinspector import add_introspection_rules
    add_introspection_rules(rules=[], patterns=['djorm_pgfulltext\.fields\.VectorField'])
except ImportError:
    pass


if django.VERSION[:2] >= (1, 7):
    # Create custom lookups for Django>= 1.7

    from django.db.models import Lookup

    def quotes(wordlist):
        return ["%s" % adapt(x.replace("\\", "")) for x in wordlist]

    def startswith(wordlist):
        return [x + ":*" for x in quotes(wordlist)]

    def negative(wordlist):
        return ['!' + x for x in startswith(wordlist)]

    class TSConfig(object):
        def __init__(self, name):
            self.name = name

    class FullTextLookupBase(Lookup):

        def as_sql(self, qn, connection):
            lhs, lhs_params = qn.compile(self.lhs)
            rhs, rhs_params = self.process_rhs(qn, connection)

            if isinstance(rhs_params, basestring):
                rhs_params = [rhs_params]

            if type(rhs_params[0]) == TSConfig:
                ts = rhs_params.pop(0)
                ts_name = ts.name
                cmd = '%s @@ to_tsquery(%%s::regconfig, %%s)' % lhs

                rest = (ts_name, " & ".join(self.transform.__call__(rhs_params)))
            else:
                cmd = '%s @@ to_tsquery(%%s)' % lhs
                rest = (" & ".join(self.transform.__call__(rhs_params)),)

            return cmd, rest

    class FullTextLookup(FullTextLookupBase):
        """This lookup scans for exact matches in the full text index.

        You can use this object like:

            Model.objects.filter(search_field__ft='Foobar')

        will produce:

            ts_query('Foobar')

        and:

            Model.objects.filter(search_field__ft=['Foobar', 'Baz', 'Quux'])

        will produce:

            ts_query('Foobar & Baz & Quux')

        and this form:

            Model.objects.filter(
                search_field__ft=[TextSearchDictionary('french'),
                    'Le Foobar', 'Le Baz'])

        will produce SQL lookup that looks like:

            ts_query('french', '''Le Foobar'' & ''Le Baz''')

        """
        lookup_name = 'ft'

        def transform(self, *args):
            return quotes(*args)

    class FullTextLookupStartsWith(FullTextLookupBase):
        """This lookup scans for full text index entries that BEGIN with
        a given phrase, like:

            Model.objects.filter(search_field__ft=['Foobar', 'Baz', 'Quux'])

        will get translated to

            ts_query('Foobar:* & Baz:* & Quux:*')
        """
        lookup_name = 'ft_startswith'

        def transform(self, *args):
            return startswith(*args)

    class FulTextLookupNotStartsWith(FullTextLookupBase):
        """This lookup scans for full text index entries that do not begin with
        a given phrase, like:

            Model.objects.filter(search_field__ft=['Foobar', 'Baz', 'Quux'])

        will get translated to

            ts_query('!Foobar:* & !Baz:* & !Quux:*')
        """
        lookup_name = 'ft_not_startswith'

        def transform(self, *args):
            return negative(*args)

    VectorField.register_lookup(FullTextLookup)
    VectorField.register_lookup(FullTextLookupStartsWith)
    VectorField.register_lookup(FulTextLookupNotStartsWith)


djorm_pgfulltext/management/commands/update_search_field.py
"""
Update search fields.
"""
from __future__ import print_function
from django.core.management.base import BaseCommand, CommandError
from django.core.exceptions import ImproperlyConfigured
from django.db import models


class Command(BaseCommand):
    help = 'Update search fields'
    args = "appname [model]"

    def handle(self, app=None, model=None, **options):
        if not app:
            raise CommandError("You must provide an app to update search fields.")

        # check application

        try:
            app_module = models.get_app(app)
        except ImproperlyConfigured:
            raise CommandError("There is no enabled application matching '%s'." % app)

        app_models = []

        # get models

        if model:
            m = models.get_model(app, model)
            if not m:
                raise CommandError("There is no model '%s'." % model)

            app_models.append(models.get_model(app, model))
        else:
            app_models += models.get_models(app_module)

        # get models only with search managers

        app_models_for_process = [x for x in app_models if getattr(x, '_fts_manager', None)]

        if not app_models_for_process:
            raise CommandError("There is no models for processing.")

        # processing

        for m in app_models_for_process:
            print("Processing model %s..." % m, end='')
            m._fts_manager.update_search_field()
            print("Done")


djorm_pgfulltext/models.py
# -*- coding: utf-8 -*-

from itertools import repeat
from django.db import models, connections
from django.db.models.query import QuerySet

# Compatibility import and fixes section.

try:
    from django.utils.encoding import force_unicode as force_text
except ImportError:
    from django.utils.encoding import force_text

try:
    from django.db.transaction import atomic
except ImportError:
    # This encapsulates pre django 1.6 transaction
    # behavior under same abstraction as django 1.6 atomic
    # decorator. This not intends to emulate django 1.6 atomic
    # behavior, only has partially same interface for easy
    # use.
    from django.db import transaction

    class atomic(object):
        def __init__(self, using=None):
            self.using = using

        def __enter__(self):
            if not transaction.is_managed(using=self.using):
                transaction.enter_transaction_management(using=self.using)
                self.forced_managed = True
            else:
                self.forced_managed = False

        def __exit__(self, *args, **kwargs):
            try:
                if self.forced_managed:
                    transaction.commit(using=self.using)
                else:
                    transaction.commit_unless_managed(using=self.using)
            finally:
                if self.forced_managed:
                    transaction.leave_transaction_management(using=self.using)


def auto_update_search_field_handler(sender, instance, *args, **kwargs):
    instance.update_search_field()


class SearchManagerMixIn(object):
    """
    A mixin to create a Manager with a 'search' method that may do a full text search
    on the model.

    The manager is set up with a list of one or more model's fields that will be searched.
    It can be a list of field names, or a list of tuples (field_name, weight). It can also
    be None, in that case every CharField and TextField in the model will be searched.

    You can also give a 'search_field', a VectorField into where the values of the searched
    fields are copied and normalized. If you give it, the searches will be made on this
    field; if not, they will be made directly in the searched fields.

    When using search_field, if auto_update = True, Django signals will be used to
    automatically syncronize the search_field with the searched fields every time instances
    are saved. If not, you can call to 'update_search_field' method in model instances to do this.
    If search_field not used, both auto_update and update_search_field does nothing. Alternatively,
    you can create a postgresql trigger to do the syncronization at database level, see this:

    http://www.postgresql.org/docs/9.1/interactive/textsearch-features.html#TEXTSEARCH-UPDATE-TRIGGERS

    In both cases, you should create a text search index, on either the searched fields or
    the compound search_field, like explained here:

    http://www.postgresql.org/docs/9.1/interactive/textsearch-tables.html#TEXTSEARCH-TABLES-INDEX

    Finally, you can give a 'config', the Postgres text search configuration that will be used
    to normalize the search_field and the queries. How do you can create a configuration:

    http://www.postgresql.org/docs/9.1/interactive/textsearch-configuration.html

    To do all those actions in database, create a setup sql script for Django:

    https://docs.djangoproject.com/en/1.4/howto/initial-data/#providing-initial-sql-data
    """

    def __init__(self,
                 fields=None,
                 search_field='search_index',
                 config='pg_catalog.english',
                 auto_update_search_field=False):
        self.search_field = search_field
        self.default_weight = 'D'
        self.config = config
        self.auto_update_search_field = auto_update_search_field
        self._fields = fields

        super(SearchManagerMixIn, self).__init__()

    def contribute_to_class(self, cls, name):
        '''
        Called automatically by Django when setting up the model class.
        '''
        if not cls._meta.abstract:
            # Attach this manager as _fts_manager in the model class.
            if not getattr(cls, '_fts_manager', None):
                cls._fts_manager = self

            # Add 'update_search_field' instance method, that calls manager's update_search_field.
            if not getattr(cls, 'update_search_field', None):
                def update_search_field(self, using=None, config=None):
                    self._fts_manager.update_search_field(pk=self.pk, using=using, config=config)

                setattr(cls, 'update_search_field', update_search_field)

            if self.auto_update_search_field:
                models.signals.post_save.connect(auto_update_search_field_handler, sender=cls)

        super(SearchManagerMixIn, self).contribute_to_class(cls, name)

    def get_queryset(self):
        return SearchQuerySet(model=self.model, using=self._db)

    def search(self, *args, **kwargs):
        return self.get_queryset().search(*args, **kwargs)

    def update_search_field(self, pk=None, config=None, using=None):
        '''
        Update the search_field of one instance, or a list of instances, or
        all instances in the table (pk is one key, a list of keys or none).

        If there is no search_field, this function does nothing.
        '''

        if not self.search_field:
            return

        if not config:
            config = self.config

        if using is None:
            using = self.db

        connection = connections[using]
        qn = connection.ops.quote_name

        where_sql = ''
        params = []
        if pk is not None:
            if isinstance(pk, (list, tuple)):
                params = pk
            else:
                params = [pk]

            where_sql = "WHERE %s IN (%s)" % (
                qn(self.model._meta.pk.column),
                ','.join(repeat("%s", len(params)))
            )

        search_vector = self._get_search_vector(config, using)
        sql = "UPDATE %s SET %s = %s %s;" % (
            qn(self.model._meta.db_table),
            qn(self.search_field),
            search_vector,
            where_sql
        )

        with atomic():
            cursor = connection.cursor()
            cursor.execute(sql, params)

    def _find_text_fields(self):
        fields = [f for f in self.model._meta.fields
                  if isinstance(f, (models.CharField, models.TextField))]

        return [(f.name, None) for f in fields]

    def _parse_fields(self, fields):
        """
        Parse fields list into a correct format needed by this manager.
        If any field does not exist, raise ValueError.
        """

        parsed_fields = set()

        if fields is not None and isinstance(fields, (list, tuple)):
            if len(fields) > 0 and isinstance(fields[0], (list, tuple)):
                parsed_fields.update(fields)
            else:
                parsed_fields.update([(x, None) for x in fields])

            # Does not support field.attname.
            field_names = set(field.name for field in self.model._meta.fields if not field.primary_key)
            non_model_fields = set(x[0] for x in parsed_fields).difference(field_names)
            if non_model_fields:
                raise ValueError("The following fields do not exist in this"
                                 " model: {0}".format(", ".join(x for x in non_model_fields)))
        else:
            parsed_fields.update(self._find_text_fields())

        return parsed_fields

    def _get_search_vector(self, config, using, fields=None):
        if fields is None:
            vector_fields = self._parse_fields(self._fields)
        else:
            vector_fields = self._parse_fields(fields)

        search_vector = []
        for field_name, weight in vector_fields:
            search_vector.append(self._get_vector_for_field(field_name, weight, config, using))
        return ' || '.join(search_vector)

    def _get_vector_for_field(self, field_name, weight=None, config=None, using=None):
        if not weight:
            weight = self.default_weight

        if not config:
            config = self.config

        if using is None:
            using = self.db

        field = self.model._meta.get_field(field_name)

        connection = connections[using]
        qn = connection.ops.quote_name

        return "setweight(to_tsvector('%s', coalesce(%s.%s, '')), '%s')" % \
               (config, qn(self.model._meta.db_table), qn(field.column), weight)


class SearchQuerySet(QuerySet):
    @property
    def manager(self):
        return self.model._fts_manager

    @property
    def db(self):
        return self._db or self.manager.db

    def search(self, query, rank_field=None, rank_function='ts_rank', config=None,
               rank_normalization=32, raw=False, using=None, fields=None,
               headline_field=None, headline_document=None):
        '''
        Convert query with to_tsquery or plainto_tsquery, depending on raw is
        `True` or `False`, and return a QuerySet with the filter.

        If `rank_field` is not `None`, a field with this name will be added
        containing the search rank of the instances, and the queryset will be
        ordered by it. The rank_function and normalization are explained here:

        http://www.postgresql.org/docs/9.1/interactive/textsearch-controls.html#TEXTSEARCH-RANKING

        If an empty query is given, no filter is made so the QuerySet will
        return all model instances.

        If `fields` is not `None`, the filter is made with this fields instead
        of defined on a constructor of manager.

        If `headline_field` and `headline_document` is not `None`, a field with
        this `headline_field` name will be added containing the headline of the
        instances, which will be searched inside `headline_document`.

        Search headlines are explained here:
        http://www.postgresql.org/docs/9.1/static/textsearch-controls.html#TEXTSEARCH-HEADLINE
        '''

        if not config:
            config = self.manager.config

        db_alias = using if using is not None else self.db
        connection = connections[db_alias]
        qn = connection.ops.quote_name

        qs = self
        if using is not None:
            qs = qs.using(using)

        if query:
            function = "to_tsquery" if raw else "plainto_tsquery"
            ts_query = "%s('%s', '%s')" % (
                function,
                config,
                force_text(query).replace("'", "''")
            )

            full_search_field = "%s.%s" % (
                qn(self.model._meta.db_table),
                qn(self.manager.search_field)
            )

            # if fields is passed, obtain a vector expression with
            # these fields. In other case, intent use of search_field if
            # exists.
            if fields:
                search_vector = self.manager._get_search_vector(config, using, fields=fields)
            else:
                if not self.manager.search_field:
                    raise ValueError("search_field is not specified")

                search_vector = full_search_field

            where = " (%s) @@ (%s)" % (search_vector, ts_query)
            select_dict, order = {}, []

            if rank_field:
                select_dict[rank_field] = '%s(%s, %s, %d)' % (
                    rank_function,
                    search_vector,
                    ts_query,
                    rank_normalization
                )
                order = ['-%s' % (rank_field,)]

            if headline_field is not None and headline_document is not None:
                select_dict[headline_field] = "ts_headline('%s', %s, %s)" % (
                    config,
                    headline_document,
                    ts_query
                )

            qs = qs.extra(select=select_dict, where=[where], order_by=order)

        return qs


class SearchManager(SearchManagerMixIn, models.Manager):
    pass
